src/libraries/System.Runtime.Numerics/src/System/Number.BigInteger.cs

```diff
@@ -783,44 +780,22 @@ internal static bool TryFormatBigInteger(BigInteger value, ReadOnlySpan<char> fo
                 }
             }
 
-            // First convert to base 10^9.
-            int cuSrc = value._bits.Length;
-            // A quick conservative max length of base 10^9 representation
-            // A uint contributes to no more than 10/9 of 10^9 block, +1 for ceiling of division
-            int cuMax = cuSrc * (MaxPartialDigits + 1) / MaxPartialDigits + 1;
-            Debug.Assert((long)BigInteger.MaxLength * (MaxPartialDigits + 1) / MaxPartialDigits + 1 < (long)int.MaxValue); // won't overflow
-
-            uint[]? bufferToReturn = null;
-            Span<uint> base1E9Buffer = cuMax < BigIntegerCalculator.StackAllocThreshold ?
-                stackalloc uint[cuMax] :
-                (bufferToReturn = ArrayPool<uint>.Shared.Rent(cuMax));
+            // The Ratio is calculated as: log_{10^9}(2^32)
+            const double digitRatio = 1.0703288734719332;
+            Debug.Assert(BigInteger.MaxLength * digitRatio + 1 < Array.MaxLength); // won't overflow
```

Comment:
If the length doesn't need to be exact, you can use integer estimation instead, similar to what I did in `DebuggerDisplay`.

Reply:
Which part are you referring to? `NumberToBigInteger` also uses $\log_{2^{32}}(10^9)$.

https://github.com/dotnet/runtime/blob/264e39c7b3bcde1c75b20bfef60005ab059b8290/src/libraries/System.Runtime.Numerics/src/System/Number.BigInteger.cs#L562-L573

I rewrote it to use $\log_{2^{32}}(10)$ in pull request #97589, but it is essentially the same.

https://github.com/dotnet/runtime/blob/0d426df37c6ac2c1d22f952a2e359ffd4281d47b/src/libraries/System.Runtime.Numerics/src/System/Number.BigInteger.cs#L372-L379